editoast/src/converters/generate_routes.rs

```diff
@@ -0,0 +1,438 @@
+                }
+            }
+        }
+    }
+
+    fn add_directed_edge(&mut self, u: Node, v: Node, edge_type: EdgeType) {
+        self.edges.insert((u.clone(), v.clone()), edge_type);
+        self.successors.entry(u).or_default().push(v);
+    }
+
+    fn add_symmetrical_edge(&mut self, u: Node, v: Node, edge_type: EdgeType) {
+        self.add_directed_edge(u.clone(), v.clone(), edge_type.clone());
+        self.add_directed_edge(v, u, edge_type);
+    }
+
+    /* Part 3: compute the different routes */
+
+    // Computes all the routes from one TrackEndpoint to all others
+    // The routes don’t go beyond a detector or a buffer stop
+    fn one_to_all_routes(&self, start: Node) -> Vec<Route> {
+        let mut result = vec![];
+        let mut count = 0;
+        let mut pred = HashMap::new();
+        let mut queue = Vec::from([&start]);
+
+        while let Some(current) = queue.pop() {
+            if let Some(successors) = self.successors.get(current) {
+                for succ in successors {
+                    if self.valid_successor(&start, current, succ, &pred) {
+                        pred.insert(succ, current);
+                        match &succ {
+                            // All routes end at a buffer or detector and we build it
+                            Node::BufferStop(_) | Node::Detector(_) => {
+                                result.push(self.build_route(count, succ, &pred));
+                                count += 1;
+                            }
+                            Node::TrackEndpoint(_track_endpoint) => {
+                                queue.push(succ);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        result
+    }
+
+    // Can we actually use that edge in our route search
+    fn valid_successor(
+        &self,
+        start: &Node,
+        current: &Node,
+        succ: &Node,
+        pred: &HashMap<&Node, &Node>,
+    ) -> bool {
+        let edge = self
+            .edges
+            .get(&(current.clone(), succ.clone()))
+            .expect("Edge does not exist");
+        let previous_edge = pred
+            .get(current)
+            .and_then(|&p| self.edges.get(&(p.clone(), current.clone())));
+
+        let switch_u_turn = matches!(edge, EdgeType::Switch { .. })
+            && matches!(previous_edge, Some(EdgeType::Switch { .. }));
+
+        // Don’t make a U-turn on a detector
+        // -o---d>--o- The detector is only in one direction
+        //   \__<__/   The is a bypass in the opposite direction
+        // We don’t want to reach the detector through the bypass
+        let detector_u_turn = (matches!(edge, EdgeType::Track | EdgeType::ToDetector)
+            && matches!(previous_edge, Some(EdgeType::FromDetector)))
+            || (matches!(edge, EdgeType::ToDetector)
+                && matches!(previous_edge, Some(EdgeType::Track | EdgeType::ToDetector)));
```

Comment:
Shouldn't it be :
```suggestion
                && matches!(previous_edge, Some(EdgeType::Track | EdgeType::FromDetector)));
```

Reply:
:exploding_head: thanks a lot for catching this